### PR TITLE
Cli change for file based request source

### PR DESCRIPTION
**Pull request by wjuan-AFK**
Commit to add RequestSourcePlugin option to the CLI. base PR: https://github.com/envoyproxy/nighthawk/pull/560.
This is necessary to make use of the previous PRs adding the requestsourcepluginfactory via the CLI.

### CODE REVIEW

**Comment by mum4k** [comment]
@dubious90 please review and assign back to me once done.

File: test/options_test.cc [hunk 1]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+  // the unexpected (yaml) diff.
+
+  // The predicates are defined as proto maps, and these seem to re-serialize into a different
+  // order. Hence we trim the maps to contain a single entry so they don't thwart our textual
+  // comparison below.
+  EXPECT_EQ(1, cmd->mutable_failure_predicates()->erase("benchmark.http_4xx"));
+  EXPECT_EQ(1, cmd->mutable_failure_predicates()->erase("benchmark.http_5xx"));
+  EXPECT_EQ(1, cmd->mutable_failure_predicates()->erase("requestsource.upstream_rq_5xx"));
+
+  OptionsImpl options_from_proto(*cmd);
```
Comment (oschaaf) [inline test/options_test.cc:362]: Maybe leave `// TODO(#433)` here and below to make sure we don't forget to clean this up once we have a good test helper for diffing protos.

File: source/client/options_impl.cc [hunk 2]
```diff
@@ -265,7 +265,14 @@ OptionsImpl::OptionsImpl(int argc, const char* const* argv) {
       "Remote gRPC source that will deliver to-be-replayed traffic. Each worker will separately "
       "connect to this source. For example grpc://127.0.0.1:8443/.",
       false, "", "uri format", cmd);
-
+  TCLAP::ValueArg<std::string> request_source_plugin_config(
+      "", "request-source-plugin-config",
+      "Request Source plugin configuration in json or compact yaml. "
+      "Mutually exclusive with --request-source. Example (json): "
```
Comment (oschaaf) [inline source/client/options_impl.cc:271]: Is the mutual exclusivity enforced? Or will one option be prioritized over the other when both are configured?

File: ci/do_ci.sh [hunk 3]
```diff
@@ -47,7 +47,8 @@ function do_unit_test_coverage() {
 
 function do_integration_test_coverage() {
     export TEST_TARGETS="//test:python_test"
-    export COVERAGE_THRESHOLD=78.6
+    #TODO(#564): Revert this to 78.6
+    export COVERAGE_THRESHOLD=75
```
Comment (oschaaf) [inline ci/do_ci.sh:51]: This is fine, but looking at this, maybe we should leave a note here that the comparison is limited to one tenths of a percent (as that is the precision at which is gets determined by the coverage flow).

**Review by oschaaf — CHANGES_REQUESTED** [review]
Looks great to me overall, a few small code level questions/asks in review comments.

Those aside, looking at the coverage report I wonder if it would be hard to:
- test the too-large-of-a-file edge case
- cover  `OptionsListRequestSource::initOnThread()`, perhaps only to ensure it doesn't crash. If it's impossible to hit that maybe add `NOT_IMPLEMENTED_GCOVR_EXCL_LINE` to make it assert/document that.
- test bad json input as a request source config input

File: api/client/options.proto [hunk 4]
```diff
@@ -104,6 +105,7 @@ message H1ConnectionReuseStrategy {
 
 // TODO(oschaaf): Ultimately this will be a load test specification. The fact that it
 // can arrive via CLI is just a concrete detail. Change this to reflect that.
+// highest unused number is 38
```
Comment (dubious90) [inline api/client/options.proto:108]: I don't think it's conventional to add in a helper comment like this, partially because it is likely to get out of sync with future changes.  I'd suggest removing it.

File: README.md [hunk 5]
```diff
@@ -109,6 +110,13 @@ Perform a simple single warmup request (per worker) before starting
 execution. Note that this will be reflected in the counters that
 Nighthawk writes to the output. Default is false.
 
+--request-source-plugin-config <string>
+Request Source plugin configuration in json or compact yaml. Mutually
+exclusive with --request-source. Example (json):
```
Comment (dubious90) [inline README.md:115]: I believe it's also mutually exclusive with --request-options.  If we are going to document this, we should do so completely/consistently.  Should we also be updating the documentation of the old fields to indicate their mutual exclusivity?

File: api/client/options.proto [hunk 6]
```diff
@@ -148,6 +150,8 @@ message CommandLineOptions {
     // Remote gRPC source that will deliver to-be-replayed traffic. Each worker will separately
     // connect to this source.
     RequestSource request_source = 26;
+    // Selects and configures a RequestSource plugin.
```
Comment (dubious90) [inline api/client/options.proto:153]: This comment doesn't provide any extra information to the user about why and how they would use this field.

File: source/client/factories_impl.cc [hunk 7]
```diff
@@ -169,15 +169,15 @@ RequestSourceFactoryImpl::create(const Envoy::Upstream::ClusterManagerPtr& clust
     setRequestHeader(*header, option_header.header().key(), option_header.header().value());
   }
 
-  if (options_.requestSource() == "") {
-    return std::make_unique<StaticRequestSourceImpl>(std::move(header));
-  } else {
+  if (options_.requestSource() != "") {
```
Comment (dubious90) [inline source/client/factories_impl.cc:172]: pre-existing, but since we edited this code:  Can we change this to be `.empty()` instead of doing empty string equality?

Comment (dubious90) [inline source/client/factories_impl.cc:172]: I'd also wonder why we edited this code.  It seems that we took a statement that was:
```
If something is true{}
else{}
```
and made it
```
if something is not true{}
else{}
```
In general, I think not true, else statements tend to be a tad more confusing.

File: source/client/options_impl.cc [hunk 8]
```diff
@@ -496,6 +503,17 @@ OptionsImpl::OptionsImpl(int argc, const char* const* argv) {
       throw MalformedArgvException(e.what());
     }
   }
+  if (!request_source_plugin_config.getValue().empty()) {
+    try {
+      request_source_plugin_config_.emplace(envoy::config::core::v3::TypedExtensionConfig());
```
Comment (dubious90) [inline source/client/options_impl.cc:508]: This is basically just a copy of the above code, which leads me to ask:
Would it be possible / help readability if we abstracted this recipe into a private helper?  SetFieldFromJSON or something?

File: test/options_test.cc [hunk 9]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
   EXPECT_TRUE(util(*(options_from_proto.toCommandLineOptions()), *cmd));
 }
 
+// This test covers --RequestSourcePlugin, which can't be tested at the same time as --RequestSource
+// and some other options. This is the test for the inlineoptionslistplugin.
+TEST_F(OptionsImplTest, InLineOptionsListRequestSourcePlugin) {
```
Comment (dubious90) [inline test/options_test.cc:282]: We try to name tests with the action you are testing/verifying.  It should be clear what a test is accomplishing before reading the code.

File: test/options_test.cc [hunk 10]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+      "},"
+      "}"
+      "}";
+
+  std::unique_ptr<OptionsImpl> options = TestUtility::createOptionsImpl(
+      fmt::format("{} --request-source-plugin-config {} {}", client_name_, request_source_config,
+                  good_test_uri_));
+
+  // Check that our conversion to CommandLineOptionsPtr makes sense.
+  CommandLineOptionsPtr cmd = options->toCommandLineOptions();
```
Comment (dubious90) [inline test/options_test.cc:301]: Stylistically, please avoid abbreviations in variable names.

File: test/options_test.cc [hunk 11]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
   EXPECT_TRUE(util(*(options_from_proto.toCommandLineOptions()), *cmd));
 }
 
+// This test covers --RequestSourcePlugin, which can't be tested at the same time as --RequestSource
+// and some other options. This is the test for the inlineoptionslistplugin.
```
Comment (dubious90) [inline test/options_test.cc:281]: nit:  If I understand correctly, this isn't a test of inlineoptionslistplugin.  It is a test of --RequestSourcePlugin that happens to use the inlineoptionslistplugin as its test requeest source plugin.

This is important because it would be weird to have a test of inlineoptionslistplugin in this file.

File: test/options_test.cc [hunk 12]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+
+  std::unique_ptr<OptionsImpl> options = TestUtility::createOptionsImpl(
+      fmt::format("{} --request-source-plugin-config {} {}", client_name_, request_source_config,
+                  good_test_uri_));
+
+  // Check that our conversion to CommandLineOptionsPtr makes sense.
+  CommandLineOptionsPtr cmd = options->toCommandLineOptions();
+  EXPECT_TRUE(
+      util(cmd->request_source_plugin_config(), options->requestSourcePluginConfig().value()));
+  // Now we construct a new options from the proto we created above. This should result in an
```
Comment (dubious90) [inline test/options_test.cc:304]: I think you're overdescribing here.  It's also particularly confusing because the code you're describing also doesn't appear to be the code that it immediately follows this comment.

File: test/options_test.cc [hunk 13]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+  EXPECT_EQ(1, cmd->mutable_failure_predicates()->erase("benchmark.http_4xx"));
+  EXPECT_EQ(1, cmd->mutable_failure_predicates()->erase("benchmark.http_5xx"));
+  EXPECT_EQ(1, cmd->mutable_failure_predicates()->erase("requestsource.upstream_rq_5xx"));
+
+  OptionsImpl options_from_proto(*cmd);
+  std::string s1 = Envoy::MessageUtil::getYamlStringFromMessage(
+      *(options_from_proto.toCommandLineOptions()), true, true);
+  std::string s2 = Envoy::MessageUtil::getYamlStringFromMessage(*cmd, true, true);
+  EXPECT_EQ(s1, s2);
+  // For good measure, also directly test for proto equivalence, though this should be
```
Comment (dubious90) [inline test/options_test.cc:321]: If it is superfluous, it is likely (by definition) not necessary.

File: test/options_test.cc [hunk 14]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+  // superfluous.
+  EXPECT_TRUE(util(*(options_from_proto.toCommandLineOptions()), *cmd));
+}
+
+// This test covers --RequestSourcePlugin, which can't be tested at the same time as --RequestSource
+// and some other options. This is the test for the filebasedplugin.
+TEST_F(OptionsImplTest, FileBasedRequestSourcePlugin) {
```
Comment (dubious90) [inline test/options_test.cc:328]: Having now read beyond the first test, I am a little confused.  What are these tests accomplishing?  If we are testing the functionality of the new CLI option, then one test suffices.  If we are testing functionality of our different plugins, those tests should be in files related to the plugins.  (I am assuming those plugin-specific tests already exist, which makes these tests unnecessary entirely).

File: test/options_test.cc [hunk 15]
```diff
@@ -276,6 +277,141 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+      "options:[{request_method:\"1\",request_headers:[{header:{key:\"key\",value:\"value\"}}]}]"
+      "},"
+      "}"
+      "}";
+
+  std::unique_ptr<OptionsImpl> options = TestUtility::createOptionsImpl(
+      fmt::format("{} --request-source-plugin-config {} {}", client_name_, request_source_config,
+                  good_test_uri_));
+
+  // Check that our conversion to CommandLineOptionsPtr makes sense.
```
Comment (dubious90) [inline test/options_test.cc:300]: This comment doesn't seem to be helpful, as it is obvious by reading the code directly below it.

File: README.md [hunk 16]
```diff
@@ -109,6 +110,13 @@ Perform a simple single warmup request (per worker) before starting
 execution. Note that this will be reflected in the counters that
 Nighthawk writes to the output. Default is false.
 
+--request-source-plugin-config <string>
+Request Source plugin configuration in json or compact yaml. Mutually
```
Comment (dubious90) [inline README.md:114]: Request Source is not meaningful to the average user of nighthawk.  Should we point them to request source documentation?

In other words - If someone doesn't know what request source is, how would they know to use this feature?

**Review by dubious90 — CHANGES_REQUESTED** [review]

File: source/client/factories_impl.cc (hunk 7 above)
Comment (wjuan-AFK) [reply source/client/factories_impl.cc:172]: I wanted it this way because the default option is the static one, so since in a later PR I am adding an additional condition, I thought it made sense to make the non functional change here to set that up.

Comment (dubious90) [reply source/client/factories_impl.cc:172]: In the future, I think it's easier cognitively on the reviewers if you include the change in the PR where it is relevant, rather than in an unrelated PR.

If it's substantial, you could split it into its own PR, with the intent of making the future PR easier.  I don't think that would be warranted on this particular change

File: api/client/options.proto (hunk 4 above)
Comment (wjuan-AFK) [reply api/client/options.proto:108]: This seems to be relatively standard practice. It provides the convenience of not having to look through the proto to find what the highest number is. I think it is unlikely to get out of sync because if you're editing the proto, the comment is readily available where the proto is, and it only needs to be updated when you're trying to add a new field to the proto.

File: README.md (hunk 5 above)
Comment (wjuan-AFK) [reply README.md:115]: That would make sense, but --request-options doesn't actually seem to be a command line options field. I am updating --request-source to indicate their mutual exclusivity though. Currently, the only way we can get request-options seems to be via StartRequest being passed to HandleExecutionRequest.

File: source/client/options_impl.cc (hunk 2 above)
Comment (wjuan-AFK) [reply source/client/options_impl.cc:271]: The mutually exclusivity should be enforced. I added a validation in the constructor.

File: ci/do_ci.sh (hunk 3 above)
Comment (wjuan-AFK) [reply ci/do_ci.sh:51]: That makes sense thanks.

**Comment by wjuan-AFK** [comment]
> * request source config

Added these tests. The bad json test is inside options_test. The others are inside request_source_plugin_test.

File: test/options_test.cc (hunk 1 above)
Comment (wjuan-AFK) [reply test/options_test.cc:339]: Makes sense.

File: test/options_test.cc (hunk 11 above)
Comment (wjuan-AFK) [reply test/options_test.cc]: That makes sense. Thanks dubious90. I see what you mean.

File: test/options_test.cc (hunk 9 above)
Comment (wjuan-AFK) [reply test/options_test.cc]: That makes sense.

File: test/options_test.cc (hunk 10 above)
Comment (wjuan-AFK) [reply test/options_test.cc]: Sure thing.

File: test/options_test.cc (hunk 12 above)
Comment (wjuan-AFK) [reply test/options_test.cc:295]: I found this comment helpful to me, so I thought it made sense to keep it. I think I would have found these lines odd if I didn't have them.

File: test/options_test.cc (hunk 13 above)
Comment (wjuan-AFK) [reply test/options_test.cc]: Sure.

File: test/options_test.cc (hunk 14 above)
Comment (wjuan-AFK) [reply test/options_test.cc]: I'm just testing that these different parameters pass in as options successfully.

File: test/options_test.cc (hunk 15 above)
Comment (wjuan-AFK) [reply test/options_test.cc:291]: Understood.

File: test/options_test.cc [hunk 17]
```diff
@@ -272,10 +273,113 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+  // comparison below.
+  EXPECT_EQ(1, command->mutable_failure_predicates()->erase("benchmark.http_4xx"));
+  EXPECT_EQ(1, command->mutable_failure_predicates()->erase("benchmark.http_5xx"));
+  EXPECT_EQ(1, command->mutable_failure_predicates()->erase("requestsource.upstream_rq_5xx"));
+  // TODO(#433)
+  OptionsImpl options_from_proto(*command);
+  std::string s1 = Envoy::MessageUtil::getYamlStringFromMessage(
+      *(options_from_proto.toCommandLineOptions()), true, true);
+  std::string s2 = Envoy::MessageUtil::getYamlStringFromMessage(*command, true, true);
+  EXPECT_EQ(s1, s2);
```
Comment (oschaaf) [inline test/options_test.cc:310]: Consider also adding the following comparison, as I suspect the yaml/text based comparison may miss edge cases.

```c++
  EXPECT_TRUE(util(*(options_from_proto.toCommandLineOptions()), *command));
```

**Review by oschaaf — APPROVED** [review]
LGTM, thanks!
Leaving one last suggestion with respect to a test that now uses a textual yaml diff comparison as the sole means to test for equivalence, I suspect that doubling down with a true equivalence tests helps catch edge cases. (We really need https://github.com/envoyproxy/nighthawk/issues/433)

File: test/options_test.cc (hunk 17 above)
Comment (wjuan-AFK) [reply test/options_test.cc:310]: Ok. Understood. That makes sense.

Comment (wjuan-AFK) [reply test/options_test.cc:310]: @dubious90 Wanted to let you know that I was adding this back in since this reasoning makes sense to me.

File: test/options_test.cc [hunk 18]
```diff
@@ -249,7 +250,7 @@ TEST_F(OptionsImplTest, AlmostAll) {
   EXPECT_TRUE(util(cmd->stats_sinks(0), options->statsSinks()[0]));
   EXPECT_TRUE(util(cmd->stats_sinks(1), options->statsSinks()[1]));
   EXPECT_EQ(cmd->latency_response_header_name().value(), options->responseHeaderWithLatencyInput());
-
+  // TODO(#433)
```
Comment (dubious90) [inline test/options_test.cc:253]: Two nit asks on this set of TODOs:
1.  Can we add a oneliner, explaining what the action to do here specifically is
2.  Rather than repeating it on every test here, we could have in the comment something like "Here and below".  Then we only need to include the TODO once per relevant file

File: test/options_test.cc (hunk 12 above)
Comment (dubious90) [reply test/options_test.cc:294]: Sorry -- my previous code comment wasn't particularly helpful/directive.  I agree that comments here would be helpful, but I'm not finding this comment to clarify the code for me, so let me re-explain better:
In general, comments are most helpful when we explain the why, rather than the what.  This comment is actually doing a little bit of both.  One of the reasons why I think you're finding the need to explain "what" is that the comments aren't actually positionally with the code that they're explaining.

Rather than explaining a whole series of actions in one comment, can we split up the comments to be directly tied to their code sections?

File: test/options_test.cc [hunk 19]
```diff
@@ -272,10 +273,115 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+  // order. Hence we trim the maps to contain a single entry so they don't thwart our textual
+  // comparison below.
+  EXPECT_EQ(1, command->mutable_failure_predicates()->erase("benchmark.http_4xx"));
+  EXPECT_EQ(1, command->mutable_failure_predicates()->erase("benchmark.http_5xx"));
+  EXPECT_EQ(1, command->mutable_failure_predicates()->erase("requestsource.upstream_rq_5xx"));
+  // TODO(#433)
+  OptionsImpl options_from_proto(*command);
+  std::string s1 = Envoy::MessageUtil::getYamlStringFromMessage(
+      *(options_from_proto.toCommandLineOptions()), true, true);
+  std::string s2 = Envoy::MessageUtil::getYamlStringFromMessage(*command, true, true);
```
Comment (dubious90) [inline test/options_test.cc:309]: One thing that might make this easier to parse is if we use more descriptive names, rather than s1 and s2.  Maybe expected_options_yaml and actual_options_yaml?  Or whatever names you think would be accurate/helpful.

File: test/options_test.cc [hunk 20]
```diff
@@ -272,10 +273,115 @@ TEST_F(OptionsImplTest, RequestSource) {
...
+  EXPECT_EQ(s1, s2);
+  // Additional comparison to avoid edge cases missed.
+  EXPECT_TRUE(util(*(options_from_proto.toCommandLineOptions()), *command));
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    HappyPathRequestSourceConfigJsonSuccessfullyTranslatesIntoOptions,
+    RequestSourcePluginTestFixture,
+    ::testing::ValuesIn(std::vector<std::string>{
```
Comment (dubious90) [inline test/options_test.cc:318]: I'm having a really hard time parsing this.  There are two things that might help here:
1. C++ has a formatted string option that will make this much easier to read, rather than using \" so much. You can do R"( string that contains " with no problems)". https://abseil.io/tips/64
2. (optional) I _think_ this is just one string, but it's difficult to tell.  Can we abstract the actual textprotos into variables?

**Review by dubious90 — COMMENTED** [review]
LGTM modulo a few commenting comments.  Will continue tracking the two conversations there, but otherwise reassigning to mumak, who will approve.

File: test/options_test.cc (hunk 18 above)
Comment (wjuan-AFK) [reply test/options_test.cc:253]: 1. sounds good to me. I feel like tagging the comparisons which need to be replaced is helpful, but I can remove the other ones if you insist.

File: test/options_test.cc (hunk 19 above)
Comment (wjuan-AFK) [reply test/options_test.cc:309]: For sure.

File: test/options_test.cc (hunk 20 above)
Comment (wjuan-AFK) [reply test/options_test.cc]: I'm not sure that 1 works for me here. I might be missing something, but it seems like there are issues with the newlines. At least it seems like using R"( makes you keep the newlines if you space them into separate lines. I can make each line a separate raw string though, I'm not sure that this makes it more readable, but maybe you'll find this preferable?
2. I've kind of tried to do this. Hope that this approach looks ok.

**Review by mum4k — APPROVED** [review]
Looks good, no actionable comments from my end.

**Comment by mum4k** [comment]
@dubious90 this is good to go.

**Review by dubious90 — APPROVED** [review]
